source/sessions/driver-sessions.rst

```diff
@@ -954,10 +954,24 @@ topologies.  It MUST NOT be run against a standalone server.
     * Capture the command sent to the server
     * Assert that the command sent to the server does not have an ``lsid`` field
 
-6. At the end of every individual functional test of the driver, there SHOULD be an assertion that
+6. Client-side cursor that exhausts the results on the initial query immediately returns implicit session
+to pool.
+    * Insert one document into a collection
+    * Execute a find operation on the collection
+    * Assert that the implicit session is returned to the pool prior to iterating the cursor
```

Comment:
I anticipate fractious cavils over this.
```python
cursor = collection.find()
# The session should be returned *prior* to this? What if no session was yet allocated?
# Or if the cursor sends no "find" command until iteration begins?
for document in cursor:
    pass
```
If what you're really worried about is findOne, then perhaps this test is only for drivers with a findOne method, and the test is simply that the implicit session used for findOne is returned.

Reply:
This test won't work in the .NET driver.

The implicit session is reference counted. The cursor returned by the find operation retains a reference count on the session, so the session is NOT returned to the session pool when find returns, but rather when the cursor is exhausted.

Reply:
Maybe what you are saying is that if all the matching documents (perhaps only one in this test?) are returned in the initial batch that the cursor does not need to retain a reference count on the session (implicit or otherwise) because it already knows it will never have to execute a getMore?

Reply:
@ajdavis could change the test to add two documents, and assert that the session is returned to the pool after iterating past the first one but before the cursor is closed.

Reply:
I had to read between the lines to realize that the important thing about this test is that the entire results fit in the first batch (and that therefore the cursorId would be 0, no getMore would ever be executed, and the cursor would not need to retain a reference to the session). Perhaps the language could be more explicit.

Thinking about it some more I don't see how I could write this test for the .NET driver. By the time Find has returned, the implicit session no longer exists so not sure what to assert against. We could assert that the cursor's session property is null but that's not exactly the same thing.

Reply:
It requires access to the server session pool to get a count of checked-out sessions.  This same capability is required by all three new tests.  I had to add an internal method to the server session pool class in the java driver, as well as an internal method in MongoClient to access its server session pool

Reply:
I wondered if that's what you were thinking. That seems brittle. For example, if the server session pool is empty when the test starts the count will be one greater afterwards.

It also can't work at all when tests are run in parallel, which is one of our goals to speed up the .NET driver test suite.

Reply:
I was thinking to assert that the checked-out count is zero, not the total available.  

Agreed that parallel tests make this harder.  I don't have a solution for that.  But since we've been living with non-parallel tests for years, I'm not that worried about it and think this is worthwhile until then.

Reply:
I was right about the fractious cavils! I agree with your idea to insert two documents, start iterating the cursor, and assert that the session has been returned to the pool after the first document is retrieved from the cursor. A driver can use APM to determine the lsid used with "find" and then check that the lsid is back in the pool.

If there's no API for inspecting the session pool, you could do:
```
create a new client with an empty session pool
register an APM command-started listener
cursor = collection.find()
doc0 = cursor.next()
write down the lsid used with "find"
collection.insert_one({})
assert the same lsid was used for "insert"
```
That proves the lsid was returned to the pool before we called "cursor.next" again.

Reply:
Excellent idea.  I incorporated it into the same test by listing two different ways to assert that the session has been returned to the pool.

Reply:
> I incorporated it into the same test by listing two different ways to assert that the session has been returned to the pool.

Did you add this? I don't see it.

Reply:
Pushing to the branch helps... Thanks Shane.